athenahealth/httpclient.go

```diff
@@ -172,8 +178,15 @@ func (h *HTTPClient) request(ctx context.Context, method, path string, body io.R
 		h.requestLock.Unlock()
 
 		if errors.Is(err, ratelimiter.ErrRateExceeded) {
-			time.Sleep(retryAfter)
-			return h.request(ctx, method, path, body, headers, out)
+			select {
+			case <-ctx.Done():
+				return nil, fmt.Errorf("waiting for rate limit timeout: %w", ctx.Err())
```

Comment:
wonder if there's a less ambiguous way to say this (like `"timed out waiting for rate limit retry interval"?) but totally fine as-is

Reply:
In the current state, the full error would read something along the lines of:
`waiting for rate limit timeout: context deadline exceeded`
OR
`waiting for rate limit timeout: context canceled`
We can't assume that it was a timeout event as it's possible the context was canceled and not a deadline

I like the verbiage "retry interval" better so we can switch to
"waiting for rate limit retry interval"
which would give us:
`waiting for rate limit retry interval: context deadline exceeded`
OR
`waiting for rate limit retry interval: context canceled`

Reply:
Ah yup because `Done()` would catch cancellation as well as context timeouts 👍